[FEATURE] addAddonsToProject for blueprints

elwayman02 opened the pull request:
This adds the `addAddonsToProject` method to mirror the same aggregate functionality provided by `addPackagesToProject` and `addBowerPackagesToProject`
Additionally, refactors `addAddonToProject` to leverage the new method
Also adds `_writeStatusToUI` to `addBowerPackagesToProject` to bring it to parity with the npm/addon versions

Here's an example blueprint you can use to test `addAddonsToProject`:

```
module.exports = {
  description: 'Set up additional packages for Ember apps',
  normalizeEntityName: function() {}, // no-op since we're just adding dependencies

  afterInstall: function(options) {
    return this.addAddonsToProject({
      packages: [
        'ember-cli-blanket',
        'ember-cli-sass-pods',
        'ember-cpm',
        'ember-feature-flags',
        'ember-metrics',
        'ember-moment'
      ]
    });
  }
};
```

I chose several common addons that have various blueprint actions in their afterInstall so that it's clear each blueprint is being executed as expected. Additionally, you may pass objects in the `packages` array to indicate version compatibility or include it in the string itself.

```
module.exports = {
  description: 'Set up additional packages for Ember apps',
  normalizeEntityName: function() {}, // no-op since we're just adding dependencies

  afterInstall: function(options) {
    return this.addAddonsToProject({
      packages: [
        'ember-cli-blanket',
        'ember-cli-sass-pods',
        'ember-cpm',
        { name: 'ember-feature-flags', target: '1.1.0' },
        'ember-metrics',
        'ember-moment@3.4.0'
      ]
    });
  }
};
```

elwayman02 commented:
<img width="676" alt="1__bash" src="https://cloud.githubusercontent.com/assets/1325482/9836241/7272af6e-59c6-11e5-97be-ff3d3155c1cd.png">

stefanpenner commented:
SG

@mike-north got cycles to review?

elwayman02 commented:
Failing tests are from master afaict

stefanpenner commented:
> Failing tests are from master afaict

ya, i think we are getting trolled by something upstream https://github.com/npm/fstream/issues/46

mike-north commented:
this needs at the very least a test case that touches on the difference between `source` and `name`. We ran into some annoying problems with an incomplete `addBowerPackageToProject` implementation, and it has been quite a headache to fix and support

elwayman02 commented:
@mike-north what is `source`? As far as I can tell, that's not even a parameter currently supported by the existing `addAddonToProject`. I'm not sure that's in scope for this feature, but I'm happy to address it to prevent future headaches if you can explain to me what I'm looking for.

elwayman02 commented:
`addPackagesToProject` also does not support a `source` param, for the record.

mike-north commented:
Ok, upon further review, looks like npm doesn't support source to customize the name of a library in the node_modules folder [like bower allows](https://github.com/bower/endpoint-parser).

mike-north commented on lib/models/blueprint.js:
```diff
@@ -1046,29 +1049,57 @@ Blueprint.prototype.addBowerPackagesToProject = function(packages, installOption
+    throw new SilentError('You must provide package to addAddonsToProject');
+  }
+
+  var install = (packages.length > 1) ? 'install addons' : 'install addon';
```
confusing variable name

elwayman02 replied:
Agreed, this is the existing pattern from addPackagesToProject. I can refactor all usages to something else.

mike-north commented on tests/unit/models/blueprint-test.js:
```diff
@@ -1046,6 +1046,38 @@ describe('Blueprint', function() {
+      };
+
+      blueprint.addAddonToProject({name: 'foo-bar', target: '^123.1.12'});
+    });
```
Please add test coverage for a git tag or wildcard target

elwayman02 replied:
@mike-north I'm not sure what that would accomplish.  I could literally pass *anything* in target and the result would be the same.

For example: 
```
it('passes a packages array with target for addAddonsToProject', function() {
      blueprint.addAddonsToProject = function(options) {
        expect(options.packages).to.deep.equal([{name: 'foo-bar', target: '^123.1.12'}]);
      };

      blueprint.addAddonToProject({name: 'foo-bar', target: 'banana pudding pops'});
    });
```

addAddonToProject/addAddonsToProject aren't responsible for sanitizing the inputs, any more than addPackage(s)ToProject. This is something that is (or should be) handled in a more centralized place downstream, probably in `addon-install` or add`npm-task`.

elwayman02 commented:
@stefanpenner @mike-north Please let me know if there are any remaining unaddressed issues for this PR or if we can get this merged, thanks! I just rebased with master again, so it should be up-to-date now.

elwayman02 commented:
I've created an issue for Mike's comment regarding test coverage of targets, since that shouldn't be handled as part of this PR.

rwjblue commented on lib/models/blueprint.js:
```diff
@@ -892,7 +892,7 @@ Blueprint.prototype.addPackageToProject = function(packageName, target) {
 Blueprint.prototype.addPackagesToProject = function(packages) {
   var task         = this.taskFor('npm-install');
-  var install      = (packages.length > 1) ? 'install packages' : 'install package';
+  var installText      = (packages.length > 1) ? 'install packages' : 'install package';
```
Can you fix the indentation here?  I personally would prefer a single space, but if we are going to keep the alignment the `=` should be lined up...

elwayman02 replied:
I'd also prefer a single space...will just fix to that.

rwjblue commented on lib/models/blueprint.js:
```diff
@@ -949,7 +949,7 @@ Blueprint.prototype.removePackageFromProject = function(packageName) {
 Blueprint.prototype.removePackagesFromProject = function(packages) {
   var task         = this.taskFor('npm-uninstall');
-  var install      = (packages.length > 1) ? 'uninstall packages' : 'uninstall package';
+  var installText      = (packages.length > 1) ? 'uninstall packages' : 'uninstall package';
```
indentation tweaks

rwjblue commented on lib/models/blueprint.js:
```diff
@@ -1058,29 +1061,57 @@ Blueprint.prototype.addBowerPackagesToProject = function(packages, installOption
+    blueprintOptions: options.blueprintOptions || {}
+  };
+
+  var packages = options['packages'];
```
Can you swap this to `options.packages` instead?

rwjblue commented on lib/models/blueprint.js:
```diff
@@ -1058,29 +1061,57 @@ Blueprint.prototype.addBowerPackagesToProject = function(packages, installOption
+*/
+Blueprint.prototype.addAddonsToProject = function (options) {
+  var taskOptions = {
+    'packages': [],
```
Why does `packages` need to be quoted here (might have been that way before, but seems slightly odd)?

elwayman02 replied:
Yea, was following an existing pattern. I don't understand why it was done this way, my initial guess was that someone got confused between `package` and `packages` as to which was a reserved keyword in JS?

rwjblue commented:
I left a few nit-picky comments, but this looks good to me.  I believe that the test failures will be addressed by a rebase...

elwayman02 commented:
@rwjblue addressed comments and rebased

elwayman02 commented:
Running the tests on my local box times out on a seemingly unrelated `ember new` test. All other tests seem to be passing except that one, but if I run on master they all succeed. Trying again on my branch to see if I can reproduce. Any idea what might have caused that?

elwayman02 commented:
Hmm tests complete for me locally on second attempt. Might have to re-trigger the CI builds though, if they timeout.

elwayman02 commented:
:tada:

raido commented:
Super!
